### PR TITLE
Fixed remaining sign compare warnings

### PR DESCRIPTION
**Pull request by larsewi**
Ticket: CFE-3415
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by larsewi** [comment]
There is still a sign-compare warning caused by [tests/unit/cmockery.c:989](https://github.com/cfengine/core/blob/1cb68a8c31b3ffc741ca880c27ea035638366fd7/tests/unit/cmockery.c#L989). This warning should somehow be silenced before removing `-Wno-sign-compare` from [tests/static-check/run_checks.sh:11](https://github.com/cfengine/core/blob/1cb68a8c31b3ffc741ca880c27ea035638366fd7/tests/static-check/run_checks.sh#L11). I will create a ticket for this.

File: cf-execd/cf-execd-runagent.c [hunk 1]
```diff
@@ -53,8 +53,16 @@ void HandleRunagentRequest(int conn_fd)
         int length = snprintf(error_buf, sizeof(error_buf),
                               "{\"error\": \"Failed to create file stream from the connection file descriptor: %s\"}",
                               GetErrorStr());
-        assert(length < sizeof(error_buf));
-        write(conn_fd, error_buf, MIN(sizeof(error_buf), (size_t) length));
+        if (length < 0)
+        {
+            Log(LOG_LEVEL_ERR, "Unexpected failure from snprintf while writing error message to connection file descriptor: %s",
+                GetErrorStr());
```
Comment (vpodzime) [inline cf-execd/cf-execd-runagent.c:59]: This still needs to `write()` a valid JSON with `"error": "something"`.

Comment (vpodzime) [reply cf-execd/cf-execd-runagent.c:59]: But I don't think `snprintf()` can actually ever return a value less than 0.

> If an output error is encountered, a negative value is returned.

`snprintf()` does no output.

File: cf-execd/cf-execd.c [hunk 2]
```diff
@@ -561,7 +561,14 @@ static int SetupRunagentSocket()
         ret = snprintf(sock_info.sun_path, sizeof(sock_info.sun_path) - 1,
                        "%s/"CF_EXECD_RUNAGENT_SOCKET_NAME, RUNAGENT_SOCKET_DIR);
     }
-    if (ret <= (sizeof(sock_info.sun_path) - 1))
+    if (ret < 0)
```
Comment (vpodzime) [inline cf-execd/cf-execd.c:564]: Same here.

File: cf-serverd/server_transform.c [hunk 3]
```diff
@@ -993,7 +993,7 @@ static void NewHostToOldACL(Auth *old, const char *host)
 
             /* Identify next fragment: */
             const char *d = strchr(dot, '.');
-            size_t len = d ? d - dot : strlen(dot);
+            size_t len = d ? (size_t) (d - dot) : strlen(dot);
```
Comment (vpodzime) [inline cf-serverd/server_transform.c:996]: This should be type-casted implicitly on the assignment, right?

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: cf-serverd/server_transform.c (hunk 3 above)
Comment (larsewi) [reply cf-serverd/server_transform.c:996]: I guess the compiler implicitly typecasts this to a `long int`, thus an explicit type cast to `size_t` is required.

`server_transform.c:996:30: error: operand of ?: changes signedness from ‘long int’ to ‘size_t’ {aka ‘long unsigned int’} due to unsignedness of other operand [-Werror=sign-compare]`

File: cf-execd/cf-execd-runagent.c [hunk 4]
```diff
@@ -53,7 +53,7 @@ void HandleRunagentRequest(int conn_fd)
         int length = snprintf(error_buf, sizeof(error_buf),
                               "{\"error\": \"Failed to create file stream from the connection file descriptor: %s\"}",
                               GetErrorStr());
-        assert(length < sizeof(error_buf));
+        assert((size_t) length < sizeof(error_buf));
```
Comment (vpodzime) [inline cf-execd/cf-execd-runagent.c:56]: I guess `length` could be `size_t` directly, with the implicit conversion on the assignment to it above.

**Review by vpodzime — COMMENTED** [review]
Just one last nitpick, sorry. :)

File: cf-execd/cf-execd-runagent.c (hunk 4 above)
Comment (larsewi) [reply cf-execd/cf-execd-runagent.c]: Agreed

**Comment by larsewi** [comment]
@cf-bottom Jenkins, please!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6080/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6080/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6080/

**Review by vpodzime — DISMISSED** [review]
Nice! :relieved: :tada:

File: cf-execd/cf-execd-runagent.c [hunk 5]
```diff
@@ -50,7 +50,7 @@ void HandleRunagentRequest(int conn_fd)
 
         /* SIGH: There's no dprintf() on exotics. */
         char error_buf[256];
-        int length = snprintf(error_buf, sizeof(error_buf),
+        size_t length = (size_t) snprintf(error_buf, sizeof(error_buf),
```
Comment (olehermanse) [inline cf-execd/cf-execd-runagent.c:53]: AFAIK type cast is unnecessary here.

File: cf-execd/cf-execd.c [hunk 6]
```diff
@@ -561,7 +561,7 @@ static int SetupRunagentSocket()
         ret = snprintf(sock_info.sun_path, sizeof(sock_info.sun_path) - 1,
                        "%s/"CF_EXECD_RUNAGENT_SOCKET_NAME, RUNAGENT_SOCKET_DIR);
     }
-    if (ret <= (sizeof(sock_info.sun_path) - 1))
+    if ((size_t) ret <= (sizeof(sock_info.sun_path) - 1))
```
Comment (olehermanse) [inline cf-execd/cf-execd.c:564]: Can `ret < 0`? Could add an assert or comment about this.

**Review by olehermanse — DISMISSED** [review]
LGTM, some small comments.

File: cf-execd/cf-execd-runagent.c (hunk 5 above)
Comment (larsewi) [reply cf-execd/cf-execd-runagent.c:53]: I think you're right !

File: cf-execd/cf-execd.c (hunk 6 above)
Comment (larsewi) [reply cf-execd/cf-execd.c:564]: Good idea !

Comment (vpodzime) [reply cf-execd/cf-execd.c]: > Can `ret < 0`?

Not really. But `assert()` is OK, I guess.

**Review by olehermanse — APPROVED** [review]

**Review by vpodzime — APPROVED** [review]
ACK

**Pull request merged**